typing/rec_check.ml

```diff
@@ -18,145 +18,161 @@ open Types
+    (** A [Guard] context returns the value as a member of a data structure,
+        for example a variant constructor or record. The value can safely be
+        defined mutually-recursively with their context, for example in
+        [let rec li = 1 :: li].
+        When these subexpressions participate in a cyclic definition,
+        this definition is productive/guarded.
+
+        The [Guard] mode is also used when a value is not dereferenced,
+        it is returned by a sub-expression, but the result of this
+        sub-expression is discarded instead of being returned.
+        For example, the subterm [?] is in a [Guard] context
+        in [let _ = ? in e] and in [?; e].
+        When these subexpressions participate in a cyclic definition,
+        they cannot create a self-loop.
+    *)
+
+    | Delay
+    (** A [Delay] context can be fully evaluated without evaluting its argument,
+        which will only be needed at a later point of program execution. For
+        example, [fun x -> ?] or [lazy ?] are [Delay] contexts. *)
+
+    | Unused
+    (** [Unused] is for subexpressions that are not used at all during
+       the evaluation of the whole program. This is the mode of
+       a variable in an expression in which it does not occur. *)
+
+  (* Returns the most conservative mode of the two arguments.
+
+     Dereference < Return < Guard < Delay < Unused
+
+     In judgments we write (m + m') for (join m m').
+  *)
+  let join m m' =
+    match m, m' with
+    | Dereference, _
+    | _, Dereference -> Dereference
+    | Return, _
+    | _, Return -> Return
+    | Guard, _
+    | _, Guard -> Guard
+    | Delay, _
+    | _, Delay -> Delay
+    | _ -> Unused
+
+  (* If x is used with the mode m in e[x], and e[x] is used with mode
+     m' in e'[e[x]], then x is used with mode m'[m] in e'[e[x]].
+
+     Return is neutral for composition: m[Return] = m = Return[m].
+  *)
+  let compos m' m = match m', m with
+    | Unused, _ | _, Unused -> Unused
+    | Dereference, _ -> Dereference
+    | Delay, _ -> Delay
+    | Guard, Return -> Guard
+    | Guard, ((Dereference | Guard | Delay) as m) -> m
+    | Return, Return -> Return
+    | Return, ((Dereference | Guard | Delay) as m) -> m
 end
 
-open Rec_context
+type mode = Mode.t = Dereference | Return | Guard | Delay | Unused
+
+module Env :
+sig
+  type t
+
+  val single : Ident.t -> Mode.t -> t
+  (** Create an environment with a single identifier used with a given mode.
+  *)
+
+  val empty : t
+  (** An environment with no used identifiers. *)
+
+  val find : Ident.t -> t -> Mode.t
+  (** Find the mode of an indentifier in an environment.  The default mode is
+      Unused. *)
+
+  val unguarded : t -> Ident.t list -> Ident.t list
+  (** unguarded e l: the list of all identifiers in l that are dereferenced or
+      returned in the environment e. *)
+
+  val dependent : t -> Ident.t list -> Ident.t list
+  (** unguarded e l: the list of all identifiers in e that are used in e. *)
```

Comment:
Two things to fix in this comment: it should say "dependent", not "unguarded", and it should say "in l", not "in e"